File: samples/cpp/embedded-speech/samples/speech_translation_samples.cpp
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See https://aka.ms/csspeech/license for the full license information.
//

#include <iostream>
#include <thread>
#include <speechapi_cxx.h>

using namespace Microsoft::CognitiveServices::Speech;
using namespace Microsoft::CognitiveServices::Speech::Audio;
using namespace Microsoft::CognitiveServices::Speech::Translation;

extern std::shared_ptr<EmbeddedSpeechConfig> CreateEmbeddedSpeechConfig();
extern std::shared_ptr<HybridSpeechConfig> CreateHybridSpeechConfig();


// Lists available embedded speech translation models.
void ListEmbeddedSpeechTranslationModels()
{
    // Creates an instance of an embedded speech config.
    auto speechConfig = CreateEmbeddedSpeechConfig();
    if (!speechConfig)
    {
        return;
    }

    // Gets a list of models.
    auto models = speechConfig->GetSpeechTranslationModels();

    if (!models.empty())
    {
        std::cout << "Models found [" << models.size() << "]:" << std::endl;
        for (const auto& model : models)
        {
            std::cout << std::endl;
            std::cout << model->Name << std::endl;
            std::cout << " Source language(s) [" << model->SourceLanguages.size() << "]: ";
            for (const auto& sourceLang : model->SourceLanguages)
            {
                std::cout << sourceLang << " ";
            }
            std::cout << std::endl;
            std::cout << " Target language(s) [" << model->TargetLanguages.size() << "]: ";
            for (const auto& targetLang : model->TargetLanguages)
            {
                std::cout << targetLang << " ";
            }
            std::cout << std::endl;
            std::cout << " Path: " << model->Path << std::endl;
        }
    }
    else
    {
        std::cerr << "No models found. Either the path is not valid or the format of model(s) is unknown." << std::endl;
    }
}


void TranslateSpeech(std::shared_ptr<TranslationRecognizer> recognizer)
{
    // Subscribes to events.
    recognizer->Recognizing += [](const TranslationRecognitionEventArgs& e)
    {
        // Intermediate result (hypothesis).
        // Note that embedded "many-to-1" translation models support only one
        // target language (the model native output language). For example, a
        // "Many-to-English" model generates only output in English.
        // At the moment embedded translation cannot provide transcription of
        // the source language.
        if (e.Result->Reason == ResultReason::TranslatingSpeech)
        {
            // Source (input) language identification is enabled when TranslationRecognizer
            // is created with an AutoDetectSourceLanguageConfig argument.
            // In case the model does not support this functionality or the language cannot
            // be identified, the result Language is "Unknown".
            auto sourceLangResult = AutoDetectSourceLanguageResult::FromResult(e.Result);
            const auto& sourceLang = sourceLangResult->Language;

            for (const auto& translation : e.Result->Translations)
            {
                auto targetLang = translation.first;
                auto outputText = translation.second;
                std::cout << "Translating [" << sourceLang << " -> " << targetLang << "]: " << outputText << std::endl;
            }
        }
    };

    recognizer->Recognized += [](const TranslationRecognitionEventArgs& e)
    {
        // Final result. May differ from the last intermediate result.
        if (e.Result->Reason == ResultReason::TranslatedSpeech)
        {
            auto sourceLangResult = AutoDetectSourceLanguageResult::FromResult(e.Result);
            const auto& sourceLang = sourceLangResult->Language;

            for (const auto& translation : e.Result->Translations)
            {
                auto targetLang = translation.first;
                auto outputText = translation.second;
                std::cout << "TRANSLATED [" << sourceLang << " -> " << targetLang << "]: " << outputText << std::endl;
            }
        }
        else if (e.Result->Reason == ResultReason::NoMatch)
        {
            // NoMatch occurs when no speech phrase was recognized.
            auto reason = NoMatchDetails::FromResult(e.Result)->Reason;
            std::cout << "NO MATCH: Reason=";
            switch (reason)
            {
            case NoMatchReason::NotRecognized:
                // Input audio was not silent but contained no recognizable speech.
                std::cout << "NotRecognized" << std::endl;
                break;
            case NoMatchReason::InitialSilenceTimeout:
                // Input audio was silent and the (initial) silence timeout expired.
                // In continuous recognition this can happen multiple times during
                // a session, not just at the very beginning.
                std::cout << "InitialSilenceTimeout" << std::endl;
                break;
            default:
                // Other reasons are not supported in embedded speech at the moment.
                std::cout << int(reason) << std::endl;
                break;
            }
        }
    };

    recognizer->Canceled += [](const TranslationRecognitionCanceledEventArgs& e)
    {
        switch (e.Reason)
        {
        case CancellationReason::EndOfStream:
            // Input stream was closed or the end of an input file was reached.
            std::cout << "CANCELED: EndOfStream" << std::endl;
            break;

        case CancellationReason::Error:
            // NOTE: In case of an error, do not use the same recognizer for recognition anymore.
            std::cerr << "CANCELED: ErrorCode=" << int(e.ErrorCode) << std::endl;
            std::cerr << "CANCELED: ErrorDetails=\"" << e.ErrorDetails << "\"" << std::endl;
            break;

        default:
            std::cout << "CANCELED: Reason=" << int(e.Reason) << std::endl;
            break;
        }
    };

    recognizer->SessionStarted += [](const SessionEventArgs& e)
    {
        UNUSED(e);
        std::cout << "Session started." << std::endl;
    };

    recognizer->SessionStopped += [](const SessionEventArgs& e)
    {
        UNUSED(e);
        std::cout << "Session stopped." << std::endl;
    };

    // The following lines run continuous recognition that listens for speech
    // in input audio and generates results until stopped. To run recognition
    // only once, replace this code block with
    //
    // auto result = recognizer->RecognizeOnceAsync().get();
    //
    // and optionally check this returned result for the outcome instead of
    // doing it in event handlers.

    // Starts continuous recognition.
    recognizer->StartContinuousRecognitionAsync().get();

    std::cout << "Say something (press Enter to stop)..." << std::endl;
    std::cin.get();

    // Stops recognition.
    recognizer->StopContinuousRecognitionAsync().get();
}


// Translates speech using embedded speech config and the system default microphone device.
void EmbeddedSpeechTranslationFromMicrophone()
{
    auto speechConfig = CreateEmbeddedSpeechConfig();
    auto sourceLangConfig = AutoDetectSourceLanguageConfig::FromOpenRange(); // optional, for input language identification
    auto audioConfig = AudioConfig::FromDefaultMicrophoneInput();

    auto recognizer = TranslationRecognizer::FromConfig(speechConfig, sourceLangConfig, audioConfig);
    TranslateSpeech(recognizer);
}

File: samples/python/embedded-speech/speech_translation_samples.py
#!/usr/bin/env python
# coding: utf-8

# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE.md file in the project root for full license information.

"""
Embedded (offline) speech translation samples for the Speech SDK for Python.

The source and target languages of an embedded translation model are fixed by the
model itself, so they are not configured in code. Use ``get_speech_translation_models()``
to inspect what each model supports.
"""

import threading

import azure.cognitiveservices.speech as speechsdk

import settings


def embedded_translation_list_models():
    """List the embedded speech translation models available at the configured model path."""
    if not settings.has_speech_translation_model_path():
        return
    config = settings.create_embedded_speech_config()
    if config is None:
        return

    models = config.get_speech_translation_models()
    if models:
        print("Embedded speech translation models found:")
        for model in models:
            print("Name:    {}".format(model.name))
            print(" Source languages: {}".format(", ".join(model.source_languages)))
            print(" Target languages: {}".format(", ".join(model.target_languages)))
            print(" Path:    {}".format(model.path))
            print(" Version: {}".format(model.version))
    else:
        print("No embedded speech translation models found. "
              "Check the model path in settings.py.")


def _print_translations(result):
    for target_language, translation in result.translations.items():
        print("TRANSLATED into '{}': {}".format(target_language, translation))


def embedded_translation_once_from_microphone():
    """Translate speech once from the default microphone using an embedded model."""
    if not settings.has_speech_translation_model():
        return
    config = settings.create_embedded_speech_config()
    if config is None:
        return

    audio_config = speechsdk.audio.AudioConfig(use_default_microphone=True)
    recognizer = speechsdk.translation.TranslationRecognizer(
        translation_config=config, audio_config=audio_config)

    print("Speak into your microphone.")
    result = recognizer.recognize_once()

    if result.reason == speechsdk.ResultReason.TranslatedSpeech:
        _print_translations(result)
    elif result.reason == speechsdk.ResultReason.NoMatch:
        print("NOMATCH: Speech could not be recognized.")
    elif result.reason == speechsdk.ResultReason.Canceled:
        cancellation = result.cancellation_details
        print("CANCELED: Reason={}".format(cancellation.reason))
        if cancellation.reason == speechsdk.CancellationReason.Error:
            print("CANCELED: ErrorCode={}".format(cancellation.error_code))
            print("CANCELED: ErrorDetails={}".format(cancellation.error_details))


def embedded_translation_continuous_from_microphone():
    """Translate speech continuously from the default microphone using an embedded model."""
    if not settings.has_speech_translation_model():
        return
    config = settings.create_embedded_speech_config()
    if config is None:
        return

    audio_config = speechsdk.audio.AudioConfig(use_default_microphone=True)
    recognizer = speechsdk.translation.TranslationRecognizer(
        translation_config=config, audio_config=audio_config)

    done = threading.Event()

    def recognized_cb(evt: speechsdk.translation.TranslationRecognitionEventArgs):
        if evt.result.reason == speechsdk.ResultReason.TranslatedSpeech:
            _print_translations(evt.result)
        elif evt.result.reason == speechsdk.ResultReason.NoMatch:
            print("NOMATCH: Speech could not be recognized.")

    def canceled_cb(evt: speechsdk.translation.TranslationRecognitionCanceledEventArgs):
        print("CANCELED: Reason={}".format(evt.reason))
        if evt.reason == speechsdk.CancellationReason.Error:
            print("CANCELED: ErrorDetails={}".format(evt.error_details))
        done.set()

    def stop_cb(evt: speechsdk.SessionEventArgs):
        print("Session stopped.")
        done.set()

    recognizer.recognized.connect(recognized_cb)
    recognizer.canceled.connect(canceled_cb)
    recognizer.session_stopped.connect(stop_cb)

    print("Speak into your microphone. Press Enter to stop.")
    recognizer.start_continuous_recognition()

    input()
    recognizer.stop_continuous_recognition()
    done.wait()
